pallets/vrf/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+		type MostRecentVrfInputGetter: GetMostRecentVrfInputs<Self::RelayBlockHash, Slot>;
+		/// Convert account to VRF key, presumably via AuthorMapping instance
+		/// TODO: maybe TryConvert and map to error because fallible conversion
+		type AccountToVrfId: Convert<Self::AccountId, AuthorityId>;
+		/// Whether or not the round changed this block
+		type RoundChanged: RoundChangedThisBlock;
+		/// Get the selected candidate accounts from staking
+		type SelectedCandidates: Get<Vec<Self::AccountId>>;
+	}
+
+	/// Current block randomness
+	/// Set in `on_initialize`, before it will contain the randomness from the last block
+	#[pallet::storage]
+	#[pallet::getter(fn current_randomness)]
+	pub type CurrentRandomness<T> = StorageValue<_, MaybeRandomness, ValueQuery>;
+
+	/// Last block randomness
+	/// Set in `on_initialize`, before it will contain the randomness from the last last block
+	#[pallet::storage]
+	#[pallet::getter(fn last_randomness)]
+	pub type LastRandomness<T> = StorageValue<_, MaybeRandomness, ValueQuery>;
+
+	/// Current set of authorities by AuthorityId
+	/// Set in `on_initialize` upon round changes
+	#[pallet::storage]
+	#[pallet::getter(fn authorities)]
+	pub type Authorities<T> = StorageValue<_, Vec<AuthorityId>, ValueQuery>;
+
+	/// Most recent relay chain block hash
+	/// Set in `on_initialize` before setting randomness
+	#[pallet::storage]
+	#[pallet::getter(fn most_recent_relay_block_hash)]
+	pub(crate) type MostRecentRelayBlockHash<T: Config> =
+		StorageValue<_, T::RelayBlockHash, ValueQuery>;
+
+	/// Most recent relay chain slot number
+	/// Set in `on_initialize` before setting randomness
+	#[pallet::storage]
+	#[pallet::getter(fn most_recent_relay_slot_number)]
+	pub(crate) type MostRecentSlotNumber<T> = StorageValue<_, Slot, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_now: BlockNumberFor<T>) -> Weight {
+			let initial_weight = Self::set_authorities_if_round_changed();
+			let (used_weight, relay_block_hash, relay_slot_number) =
+				Self::set_most_recent_vrf_inputs(initial_weight);
+			Self::set_randomness(used_weight, relay_block_hash, relay_slot_number)
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Set authorities if the round has changed
+		fn set_authorities_if_round_changed() -> Weight {
+			if T::RoundChanged::round_changed_this_block() {
+				let selected_candidates = T::SelectedCandidates::get();
+				let returned_weight: Weight = selected_candidates.len() as u64
+					* T::DbWeight::get().read
+					+ T::DbWeight::get().write;
+				Authorities::<T>::put(
+					selected_candidates
+						.into_iter()
+						.map(|x| T::AccountToVrfId::convert(x))
+						.collect::<Vec<AuthorityId>>(),
+				);
+				returned_weight
+			} else {
+				// 1 read to check if round had changed
+				T::DbWeight::get().read
+			}
+		}
+		/// Returns args for set_randomness
+		fn set_most_recent_vrf_inputs(weight: Weight) -> (Weight, Slot, T::RelayBlockHash) {
```

Comment:
Here you are passing the weight you calculated in your previous function and add it to the end write? I feel it would be cleaner if you add weights for the different functions on_initialize and maintain an argument-less function here.

Reply:
We could do that, they are necessarily executed one after the other